Codegen/test/Performance.cs

```diff
@@ -0,0 +1,232 @@
+            fileCompiler.CompileOnce();
+
+            // Generator.
+            StringBuilder destination = new StringBuilder();
+            var columnsLayout = fileCompiler.CompilationResultsForProgram.ProgramClassDocumentSnapshot.TextSourceInfo.ColumnsLayout;
+            IGenerator generator = new DefaultGenerator(fileCompiler.CompilationResultsForProgram, destination, null, null);
+
+            // Codegen iterations.
+            Console.WriteLine("FILE;ITERATION;STEP;TIME_TAKEN");
+            Dictionary<string, TimeSpan> total = new Dictionary<string, TimeSpan>();
+            for (int i = 0; i < ITERATION_COUNT; i++)
+            {
+                // Clear target StringBuilder and force recompile.
+                destination.Clear();
+                if (forceFullParsing)
+                {
+                    // Full Compile.
+                    fileCompiler = new FileCompiler(compilationProject, fileName, false);
+                    fileCompiler.CompileOnce();
+                }
+                else
+                {
+                    // Incremental compile.
+                    ITextLine newLine = new TextLineSnapshot(fileCompiler.CompilationResultsForProgram.CobolTextLines.Count, string.Empty, null);
+                    TextChangedEvent textChangedEvent = new TextChangedEvent();
+                    textChangedEvent.TextChanges.Add(new TextChange(TextChangeType.LineInserted, newLine.LineIndex, newLine));
+                    fileCompiler.CompilationResultsForProgram.UpdateTextLines(textChangedEvent);
+                    fileCompiler.CompileOnce();
+                }
+
+                // Generate and write stats.
+                generator.Generate(fileCompiler.CompilationResultsForProgram, columnsLayout);
+                foreach (var pair in generator.PerformanceReport)
+                {
+                    Console.WriteLine($"{fileName};{i+1};{pair.Key};{pair.Value.TotalMilliseconds}");
+                    if (!total.ContainsKey(pair.Key))
+                    {
+                        total.Add(pair.Key, TimeSpan.Zero);
+                    }
+                    total[pair.Key] += pair.Value;
+                }
+            }
+
+            // Aggregated stats.
+            Console.WriteLine();
+            Console.WriteLine("Time per step :");
+            foreach (var stepData in total.Select((p, i) => new {Index = i + 1, Step = p.Key, Duration = p.Value.TotalMilliseconds}).OrderByDescending(a => a.Duration))
+            {
+                Console.WriteLine($"{stepData.Step,-25}({stepData.Index}) : Total = {stepData.Duration,15} ms, Avg = {stepData.Duration / ITERATION_COUNT,15} ms");
+            }
+
+            // Total time.
+            Console.WriteLine();
+            Console.WriteLine($"Global Codegen time : Total = {total.Values.Sum(t => t.TotalMilliseconds)} ms, Avg = {total.Values.Average(t => t.TotalMilliseconds)} ms");
```

Comment:
"Total Average" is not correct.
Currently it's the total time divided by 4, because there are 4 steps.
```
Time per step :
TreeToCode               (4) : Total =      10032,8937 ms, Avg =      5016,44685 ms
Actions                  (3) : Total =        899,8611 ms, Avg =       449,93055 ms
CreateTargetDocument     (1) : Total =        300,4088 ms, Avg =        150,2044 ms
Qualifier                (2) : Total =        152,3032 ms, Avg =         76,1516 ms

Global Codegen time : Total = 11385,4668 ms, Avg = 2846,3667 ms
```

Reply:
The two last commits fix the formatting and the issues with some tests (reentrancy and OutOfMemory). Ran successfully all tests with 5 iterations on my machine.